src/arithmetic/algebraic_expression/algebraic_expression_optimization.c

```diff
@@ -112,48 +112,45 @@ static bool __AlgebraicExpression_MulOverAdd(AlgebraicExpression **root) {
 		else if((_AlgebraicExpression_IsAdditionNode(l) && !_AlgebraicExpression_IsAdditionNode(r)) ||
 				(_AlgebraicExpression_IsAdditionNode(r) && !_AlgebraicExpression_IsAdditionNode(l))) {
 
-			// Disconnect left and right children from root.
+			// disconnect left and right children from root
 			r = _AlgebraicExpression_OperationRemoveDest((*root));
 			l = _AlgebraicExpression_OperationRemoveDest((*root));
 			ASSERT(AlgebraicExpression_ChildCount(*root) == 0);
 
-			AlgebraicExpression *add = AlgebraicExpression_NewOperation(AL_EXP_ADD);
-			AlgebraicExpression *lMul = AlgebraicExpression_NewOperation(AL_EXP_MUL);
-			AlgebraicExpression *rMul = AlgebraicExpression_NewOperation(AL_EXP_MUL);
-
-			AlgebraicExpression_AddChild(add, lMul);
-			AlgebraicExpression_AddChild(add, rMul);
-
 			AlgebraicExpression *A;
 			AlgebraicExpression *B;
-			AlgebraicExpression *C;
+			AlgebraicExpression *add = AlgebraicExpression_NewOperation(AL_EXP_ADD);
 
 			if(_AlgebraicExpression_IsAdditionNode(l)) {
-				// Lefthand side is addition.
-				// (A + B) * C = (A * C) + (B * C)
-
-				A = _AlgebraicExpression_OperationRemoveSource(l);
-				B = _AlgebraicExpression_OperationRemoveDest(l);
-				C = r;
-
+				// lefthand side is addition
+				// (A + B + C) * D = (A * D) + (B * D) + (C * D)
+				B = r;
+				uint child_count = AlgebraicExpression_ChildCount(l);
+				for(uint i = 0; i < child_count; i++) {
+					A = _AlgebraicExpression_OperationRemoveDest(l);
+					AlgebraicExpression *mul = AlgebraicExpression_NewOperation(AL_EXP_MUL);
+					AlgebraicExpression_AddChild(mul, A);
```

Comment:
For consistency between the code paths, I'd put this line at 135 (to match 149-150).

Reply:
can't, must maintain order matrix multiplication isn't commutative, the order in which we introduce children to `mul` is important.